src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -8,16 +8,142 @@
+        /// throw if the queue is already completed.
+        /// </summary>
+        /// <returns>Whether or not the operation succeeded.</returns>
+        public bool TryComplete()
+        {
+            var result = _channel.Writer.TryComplete();
+            _whenCompleted.TrySetResult(true);
+            return result;
+        }
+
+        /// <summary>
+        /// Signals that no further elements will be enqueued.  All outstanding and future
+        /// Dequeue Task will be cancelled.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is already completed.</exception>
+        public void Complete()
+        {
+            if (!TryComplete())
+            {
+                throw new InvalidOperationException($"Cannot call {nameof(Complete)} when the queue is already completed.");
+            }
+        }
+
+        public void PromiseNotToEnqueue()
+            => Volatile.Write(ref _promisedNotToEnqueue, true);
+
+        /// <summary>
+        /// Gets a task whose result is the element at the head of the queue. If the queue
+        /// is empty, the returned task waits for an element to be enqueued. If <see cref="Complete"/> 
+        /// is called before an element becomes available, the returned task is completed and
+        /// <see cref="Optional{T}.HasValue"/> will be <see langword="false"/>.
+        /// </summary>
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/23582", OftenCompletesSynchronously = true)]
+        public async ValueTask<Optional<TElement>> TryDequeueAsync(CancellationToken cancellationToken)
+        {
+            while (await _channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                if (_channel.Reader.TryRead(out var element))
+                    return new Optional<TElement>(element);
+
+                // someone else may have beat us to reading the element.  try again.
+            }
+
+            // Reader is entirely done.  Definitely can't dequeue anything at this point.
+            return default;
+        }
+
+        /// <summary>
+        /// Gets a task whose result is the element at the head of the queue. If the queue
+        /// is empty, the returned task waits for an element to be enqueued. If <see cref="Complete"/> 
+        /// is called before an element becomes available, the returned task is cancelled.
+        /// </summary>
+        [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/23582", OftenCompletesSynchronously = true)]
+        public async Task<TElement> DequeueAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var result = await TryDequeueAsync(cancellationToken).ConfigureAwait(false);
+            if (result.HasValue)
+                return result.Value;
+
+            new CancellationToken(canceled: true).ThrowIfCancellationRequested();
```

Comment:
this is copied from teh original impl, but is quite problematic.  It basically violated TPL contracts around cancellation.  we should remove this and just have all the logic here work on top of Channel.